design/core_collection_operations.md

```diff
@@ -0,0 +1,175 @@
+  * The version information allows creating a new version of the collection document but only
+    updating some of the associated data products, while leaving others at their current
+    version
+  * Similar to the collections version, it also allows for easy rollbacks, and any version
+    semantics are up to the user.
+  * In the future, we could add a schema version if the schema needs to change. For now, no
+    schema version implies a schema version of 1.
+    * An altenative would just be to add a version to the product name, like `taxa_freq2`.
+
+```
+[
+    {
+        "product": "taxa_freq",
+        "version": "r207.kbase.3"
+    },
+    {
+        "product": "genome_stats",
+        "version': "r207.kbase.14"
+    }
+]
+```
+
+## Implementation
+
+### DB schema
+
+* The collections document will be essentially as represented above, with the keys / values
+  as described and a list of associated data products.
+* There will be two separate collections - `coll_all_versions` (or something) and
+  `coll_active_versions`.
+  * The former holds all versions of all collections, and the latter holds only the active
+    version of each collection.
+* Activating a version means:
+  * Pulling the collection version document from `coll_all_versions`
+  * Setting the activation date to the current date
+  * Overwriting the collection document in `coll_active_versions`.
+    * Atomic!
+* Alternatives:
+  * Tag the active version in `coll_active_version`.
+    * There's a fundamental problem with this approach in that it's non-atomic - at some point
+      either no versions will be tagged as active or > 1 will.
+    * This can be alleviated with transactions but in Arango [transactions are only reliable in
+      non-sharded environments](https://github.com/arangodb/arangodb/issues/11424).
+    * Furthermore, filtering data becomes much slower since all the inactive versions need to
+      be filtered out. Index size goes up because you need (at least) the activity state in
+      the equality part of the index. If you want to filter inactive versions, you need a
+      separate index without the activity state in the equality part of the index.
+    * Also generally more complex than the proposed schema.
+  * Have the service manage the collection version with a monotonically increasing version, like
+    `workspace_deluxe` or `sample_service`
+    * This approach is much more complex and is not necessary for the use case proposed here. It
+      also has the filtering drawback as mentioned above.
+
+### Endpoints
+
+* Only includes immediately relevant endpoints.
+
+#### List collections
+
+* For now returns all active collections based on the < 1000 collections assumption.
+
+```
+GET /collection
+```
+
+#### Save a collection
+
+* Requires a token with the `COLLECTIONS_ADMIN` (or something) `auth2` role.
+
+```
+POST /collection/{collection_id}/version/{version}
```

Comment:
how does one get the version ID for this call? Will `/collection` return a list? I'm not seeing that specified here. Alternatively we could have a `GET /collection/{collection_id}/versions`

Reply:
The user supplies the version

Reply:
how does the user query for list of what versions exist, so that they might supply it?

Reply:
They don't provide an existing version, and for now the user could just go to the arango aardvark API if they want to be sure the version they supply doesn't exist

Reply:
Resolution per out of band conversations:

In the future we should have a GET version of this endpoint so a UI can check that a version exists before trying to POST and potentially failing. This is already mentioned in the doc so no changes needed at this point. As mentioned above, for the first demo version we can just check the DB directly.